src/aggregation/bucket/range.rs

```diff
@@ -0,0 +1,546 @@
+    }
+
+    if buckets[buckets.len() - 1].to != f64::MAX {
+        converted_buckets
+            .push(f64_to_fastfield_u64(buckets[buckets.len() - 1].to, field_type)..u64::MAX);
+    }
+
+    // fill up holes in the ranges
+    let find_hole = |converted_buckets: &[Range<u64>]| {
+        for (pos, ranges) in converted_buckets.windows(2).enumerate() {
+            if ranges[0].end > ranges[1].start {
+                return Err(TantivyError::InvalidArgument(
+                    "Overlapping ranges not supported".to_string(),
+                ));
+            }
+            if ranges[0].end != ranges[1].start {
+                return Ok(Some(pos));
+            }
+        }
+        Ok(None)
+    };
+
+    while let Some(hole_pos) = find_hole(&converted_buckets)? {
+        let new_range = converted_buckets[hole_pos].end..converted_buckets[hole_pos + 1].start;
+        converted_buckets.insert(hole_pos + 1, new_range);
+    }
+
+    Ok(converted_buckets)
+}
+
+pub fn range_to_string(range: &Range<u64>, field_type: &Type) -> String {
+    let to_str = |val: u64| {
+        if val == u64::MIN || val == u64::MAX {
+            "*".to_string()
+        } else {
+            f64_from_fastfield_u64(val, field_type).to_string()
+        }
+    };
+
+    format!("{}-{}", to_str(range.start), to_str(range.end))
+}
+
+pub fn range_to_key(range: &Range<u64>, field_type: &Type) -> Key {
+    Key::Str(range_to_string(range, field_type))
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn bucket_test_extend_range_hole() {
+        let buckets = vec![(10f64..20f64).into(), (30f64..40f64).into()];
+        let buckets = extend_validate_ranges(&buckets, &Type::U64).unwrap();
+        assert_eq!(buckets[0].start, u64::MIN);
+        assert_eq!(buckets[0].end, 10);
+        assert_eq!(buckets[1].start, 10);
+        assert_eq!(buckets[1].end, 20);
+        // Added bucket to fill hole
+        assert_eq!(buckets[2].start, 20);
+        assert_eq!(buckets[2].end, 30);
+        assert_eq!(buckets[3].start, 30);
+        assert_eq!(buckets[3].end, 40);
+    }
+
+    #[test]
+    fn bucket_range_test_negative_vals() {
+        let buckets = vec![(-10f64..-1f64).into()];
+        let buckets = extend_validate_ranges(&buckets, &Type::F64)
+            .unwrap()
+            .iter()
+            .map(|bucket| range_to_string(bucket, &Type::F64))
+            .collect_vec();
+        assert_eq!(buckets[0], "*--10");
+        assert_eq!(buckets[buckets.len() - 1], "-1-*");
+    }
+    #[test]
+    fn bucket_range_test_positive_vals() {
+        let buckets = vec![(0f64..10f64).into()];
+        let buckets = extend_validate_ranges(&buckets, &Type::F64)
+            .unwrap()
+            .iter()
+            .map(|bucket| range_to_string(bucket, &Type::F64))
+            .collect_vec();
+        assert_eq!(buckets[0], "*-0");
+        assert_eq!(buckets[buckets.len() - 1], "10-*");
+    }
+
+    #[test]
+    fn range_binary_search_test() {
```

Comment:
Remove this test. (it tests nothing)

Reply:
Replace it with a test actually testing the target SegmentRangeCollector object .